File: app/src/androidTest/java/at/priv/graf/zazentimer/DuplicateSessionTest.java
package at.priv.graf.zazentimer;

import android.view.View;
import androidx.test.espresso.UiController;
import androidx.test.espresso.ViewAction;
import org.hamcrest.Matcher;
import androidx.test.ext.junit.rules.ActivityScenarioRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;
import dagger.hilt.android.testing.HiltAndroidRule;
import dagger.hilt.android.testing.HiltAndroidTest;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition;
import static androidx.test.espresso.matcher.ViewMatchers.isAssignableFrom;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.containsString;

import android.os.SystemClock;

@HiltAndroidTest
@RunWith(AndroidJUnit4.class)
@LargeTest
public class DuplicateSessionTest {

    @Rule(order = 0)
    public HiltAndroidRule hiltRule = new HiltAndroidRule(this);

    @Rule(order = 1)
    public ActivityScenarioRule<ZazenTimerActivity> activityRule =
            new ActivityScenarioRule<>(ZazenTimerActivity.class);

    @Before
    public void init() {
        hiltRule.inject();
        activityRule.getScenario().onActivity(ZazenTimerActivity::resetDatabaseForTest);
    }

    @Test
    public void testDuplicateSessionDoesNotCrash() {
        onView(withId(R.id.recycler_sessions))
                .perform(actionOnItemAtPosition(0, clickChildViewWithId(R.id.sessionOverflow)));

        // Wait for popup menu animation to complete
        SystemClock.sleep(500);

        onView(withText(R.string.menu_copy_session))
                .perform(click());

        onView(withId(R.id.recycler_sessions))
                .check(matches(isDisplayed()));
    }

    @Test
    public void testDuplicateSessionCreatesCopyWithPrefix() {
        onView(withId(R.id.recycler_sessions))
                .perform(actionOnItemAtPosition(0, clickChildViewWithId(R.id.sessionOverflow)));

        SystemClock.sleep(500);

        onView(withText(R.string.menu_copy_session))
                .perform(click());

        onView(withText(containsString("Copy of")))
                .check(matches(isDisplayed()));
    }

    private static ViewAction clickChildViewWithId(final int id) {
        return new ViewAction() {
            @Override
            public Matcher<View> getConstraints() {
                return isAssignableFrom(View.class);
            }

            @Override
            public String getDescription() {
                return "Click child view with id " + id;
            }

            @Override
            public void perform(UiController uiController, View view) {
                View child = view.findViewById(id);
                if (child != null) {
                    child.performClick();
                }
            }
        };
    }
}

File: app/src/androidTest/kotlin/at/priv/graf/zazentimer/DuplicateSessionTest.kt
package at.priv.graf.zazentimer

import androidx.test.espresso.Espresso.onIdle
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.UiDevice
import androidx.test.uiautomator.UiSelector
import at.priv.graf.zazentimer.screens.MainPage
import dagger.hilt.android.testing.HiltAndroidTest
import org.hamcrest.Matchers.containsString
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@HiltAndroidTest
@RunWith(AndroidJUnit4::class)
@LargeTest
class DuplicateSessionTest : AbstractZazenTest() {
    private lateinit var device: UiDevice

    @Before
    fun setupDatabase() {
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        var activityRef: ZazenTimerActivity? = null
        activityRule.scenario.onActivity { activityRef = it }
        activityRef?.resetDatabaseForTest()
    }

    @Test
    fun duplicateSession_doesNotCrash() {
        MainPage()
            .verifyMainScreenIsDisplayed()
            .clickSessionOverflowAtPosition(0)
        val copyText = InstrumentationRegistry.getInstrumentation().targetContext.getString(R.string.menu_copy_session)
        device.findObject(UiSelector().text(copyText)).click()
    }

    @Test
    fun duplicateSession_createsCopyWithPrefix() {
        MainPage()
            .verifyMainScreenIsDisplayed()
            .clickSessionOverflowAtPosition(0)
        val copyText = InstrumentationRegistry.getInstrumentation().targetContext.getString(R.string.menu_copy_session)
        device.findObject(UiSelector().text(copyText)).click()

        onIdle()

        onView(withText(containsString("Copy of")))
            .check(matches(isDisplayed()))
    }
}
